gobblin-core/src/main/java/gobblin/writer/commands/TeradataBufferedInserter.java

```diff
@@ -0,0 +1,201 @@
+  private static final Logger LOG = LoggerFactory.getLogger(TeradataBufferedInserter.class);
+
+  private static final String INSERT_STATEMENT_PREFIX_FORMAT = "INSERT INTO %s.%s (%s) VALUES ";
+  private static final Joiner JOINER_ON_COMMA = Joiner.on(',');
+
+  private List<JdbcEntryData> pendingInserts;
+  private List<String> columnNames;
+  private Map<Integer, Integer> columnPosSqlTypes;
+  private String insertStmtPrefix;
+  private PreparedStatement insertPstmtForFixedBatch;
+  private Retryer<Boolean> retryer;
+
+  private int batchSize;
+  private final Connection conn;
+
+  public TeradataBufferedInserter(State state, Connection conn) {
+    this.conn = conn;
+    this.batchSize = state.getPropAsInt(WRITER_JDBC_INSERT_BATCH_SIZE, DEFAULT_WRITER_JDBC_INSERT_BATCH_SIZE);
+    if (this.batchSize < 1) {
+      throw new IllegalArgumentException(WRITER_JDBC_INSERT_BATCH_SIZE + " should be a positive number");
+    }
+  }
+
+  /**
+   * Inserts entry into buffer. If current # of entries filled batch size or it overflowed the buffer, 
+   * it will call underlying JDBC to actually insert it.
+   * {@inheritDoc}
+   * @see gobblin.writer.commands.JdbcBufferedInserter#insert(java.sql.Connection, java.lang.String, java.lang.String, gobblin.converter.jdbc.JdbcEntryData)
+   */
+  @Override
+  public void insert(String databaseName, String table, JdbcEntryData jdbcEntryData) throws SQLException {
+    if (this.columnNames == null) {
+      initializeForBatch(this.conn, databaseName, table, jdbcEntryData);
+    }
+    this.pendingInserts.add(jdbcEntryData);
+
+    if (this.pendingInserts.size() == this.batchSize) {
+      insertBatch(this.insertPstmtForFixedBatch); //reuse pre-computed Preparedstatement.
+      return;
+    }
+  }
+
+  private void insertBatch(final PreparedStatement pstmt) {
+    Callable<Boolean> insertCall = new Callable<Boolean>() { //Need a Callable interface to be wrapped by Retryer.
+      @Override
+      public Boolean call() throws Exception {
+        for (JdbcEntryData pendingEntry : TeradataBufferedInserter.this.pendingInserts) {
+          int i = 1;
+          for (JdbcEntryDatum datum : pendingEntry) {
+            Object value = datum.getVal();
+            if (value != null) {
+              pstmt.setObject(i, value);
+            }
+            else {
+              // column type is needed for null value insertion
+              pstmt.setNull(i, columnPosSqlTypes.get(i));
+            }
+            i++;
+          }
+          pstmt.addBatch();
+          pstmt.clearParameters();
+        }
+        if (LOG.isDebugEnabled()) {
+          LOG.debug("Executing SQL " + pstmt);
+        }
+        int[] execStatus = pstmt.executeBatch();
+        // Check status explicitly if driver continues batch insertion upon failure 
+        for (int status : execStatus) {
+          if (status == Statement.EXECUTE_FAILED) {
+            throw new BatchUpdateException("Batch insert failed.", execStatus);
```

Comment:
If any of the statements failed in the batch we throw an exception? The ones that succeeded will be retried by the retryer. Are the inserts idempotent?